executor/linux_fork_server_executor.cpp

```diff
@@ -0,0 +1,585 @@
+/*
+ * fuzzuf
+ * Copyright (C) 2021 Ricerca Security
+ * 
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+#include "fuzzuf/executor/linux_fork_server_executor.hpp"
+#include <chrono>
+#include <cstddef>
+#include <cassert>
+#include <memory>
+#include <optional>
+#include <boost/container/static_vector.hpp>
+#include <sched.h>
+
+#include "fuzzuf/algorithms/afl/afl_option.hpp"
+#include "fuzzuf/exceptions.hpp"
+#include "fuzzuf/executor/executor.hpp"
+#include "fuzzuf/utils/workspace.hpp"
+#include "fuzzuf/utils/common.hpp"
+#include "fuzzuf/utils/which.hpp"
+#include "fuzzuf/utils/is_executable.hpp"
+#include "fuzzuf/utils/interprocess_shared_object.hpp"
+#include "fuzzuf/utils/errno_to_system_error.hpp"
+#include "fuzzuf/feedback/inplace_memory_feedback.hpp"
+#include "fuzzuf/feedback/exit_status_feedback.hpp"
+#include "fuzzuf/feedback/put_exit_reason_type.hpp"
+#include "fuzzuf/logger/logger.hpp"
+
+bool LinuxForkServerExecutor::has_setup_sighandlers = false;
+
+LinuxForkServerExecutor* LinuxForkServerExecutor::active_instance = nullptr;
+
+/**
+ * Precondition:
+ *   - A file can be created at path path_str_to_write_input.
+ * Postcondition:
+ *     - Run process below in order, with initializing members
+ *       * Configure signal handlers (Temporary workaround for non fork server mode. This will be removed in future.)
+ *       * Parsing and preprocessing of commandline arguments of PUT.
+ *       * Generate a file for sending input to  PUT.
+ *       * Configure shared memory
+ *         - afl_shm_size should indicate the size of shared memory which PUT built using AFL style cc uses.
+ *         - bb_shm_size should indicate the size of shared memory that is used to record Basic Block Coverage by PUT built using fuzzuf-cc.
+ *         - If both parameters are set to zero, it is considered as unused, and never allocate.
+ *         - In kernel, Both parameters are round up to multiple of PAGE_SIZE, then memory is allocated.
+ *       * Configure environment variables for PUT
+ *       * If fork server mode, launch fork server.
+ *       * NOTE: Executor does not take care about binding a CPU core. The owner fuzzing algorithm is responsible to it.
+ */
+LinuxForkServerExecutor::LinuxForkServerExecutor(  
```

Comment:
I think ForkServer is no longer needed in the name. This is because non fork server mode also can be handled by this executor with a dedicated channel like FdChannel. Maybe LinuxExecutor is enough?

Reply:
I agree with you.

It worth doing, but is nice to have and it takes some time to support configurable channel from caller of Executor.
We should go on next phase 3 & 4 & ... (on treating child process, coverage support) first.
So let's get back to this discussion before releasing fuzzuf-cc.
This is memorized in [roadmap](https://github.com/fuzzuf/fuzzuf/issues/52) `Phase n-1: Support forkserver and non-forkserver in LinuxForkServerExecutor.`.